### PR TITLE
Add models & routers to send push notifications

### PR DESCRIPTION
**Pull request by kevinchan159**
- Setup all models & db
    - **Apps** table which contains all the apps that we currently support for sending push notifications
    - **Users** table which contains information about all users that we can send push notifications to. Each user is tied to an App, contains a uuid which is used to uniquely identify that user for the app, and a notificationsEnabled field which is a boolean representing whether or not that user currently has enabled notifications
- Setup all basic routers including router to send push notifications
   - Routes to Create/Update Users
   - Routes to Create/Get all Apps
   - Route to send a push notification
- Add all docs to README
Screenshot:
59745453_382910759232596_6045600067543367680_n

### CODE REVIEW

File: src/repos/AppsRepo.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+const getAllApps = async (): Promise<App[]> => {
+  try {
+    return await db().find();
+  } catch (e) {
+    throw Error('Unable to fetch all apps')
+  }
+}
+
+const getOrCreateAppWithName = async (name: string): Promise<App> => {
```
Comment (alanna-zhou) [inline src/repos/AppsRepo.ts]: why can't this be combined with `createApp`

File: src/routers/SendNotificationRouter.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+const options = {
+  token: {
+    key: `./AuthKey_${APN_AUTH_KEY_ID}.p8`,
+    keyId: APN_AUTH_KEY_ID,
+    teamId: APPLE_TEAM_ID
+  },
+  production: NODE_ENV === 'production'
+}
+const apnProvider = new apn.Provider(options);
```
Comment (alanna-zhou) [inline src/routers/SendNotificationRouter.ts]: `APNProvider`?

File: src/routers/SendNotificationRouter.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+const BAD_DEVICE_TOKEN = 'BadDeviceToken';
```
Comment (alanna-zhou) [inline src/routers/SendNotificationRouter.ts]: `BAD_DEVICE_TOKEN_STR`?

File: src/routers/SendNotificationRouter.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  constructor() {
+    super('POST');
+  }
+
+  getPath(): string {
+    return '/notify/';
+  }
+
+  async content(req: Request): Promise<Object> {
+    const { apnPayload, deviceToken, uuid } = req.body;
```
Comment (alanna-zhou) [inline src/routers/SendNotificationRouter.ts]: why can't you use `uid` over `uuid`?

File: src/repos/AppsRepo.ts (hunk 1 above)
Comment (kevinchan159) [reply src/repos/AppsRepo.ts]: nice catch

File: src/routers/SendNotificationRouter.ts (hunk 4 above)
Comment (kevinchan159) [reply src/routers/SendNotificationRouter.ts]: `uuid` refers to `universally unique identifier`. its a specific type of ID that we want so i think its clearer to have `uuid`

File: src/db/DBConnection.ts [hunk 5]
```diff
@@ -1,8 +1,11 @@
 import { Connection, ConnectionOptions, createConnection } from 'typeorm';
 
+import App from '../entities/App';
+import User from '../entities/User';
+
 const isProduction = process.env.NODE_ENV === 'production';
```
Comment (Omarrasheed) [inline src/db/DBConnection.ts]: `'production'` should be moved to a constants file

File: src/server.ts [hunk 6]
```diff
@@ -1,5 +1,8 @@
 import dotenv from 'dotenv';
 import 'reflect-metadata';
+
+dotenv.config(); // establish env variables
```
Comment (Omarrasheed) [inline src/server.ts]: Y was this moved in between imports?

File: src/entities/User.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  @Column()
+  notificationsEnabled: boolean;
+
+  @ManyToOne(type => App, app => app.users, {
+    onDelete: 'CASCADE'
+  })
+  app: App;
+
+  serialize(): SerializedUser {
```
Comment (Omarrasheed) [inline src/entities/User.ts]: nitpick: Just wanted to slap u with a random nitpick

File: src/server.ts (hunk 6 above)
Comment (kevinchan159) [reply src/server.ts]: i need access to `env` variables before importing `API` so i had to call this beforehand

File: src/entities/User.ts (hunk 7 above)
Comment (kevinchan159) [reply src/entities/User.ts]: thank u omar

File: .gitignore [hunk 8]
```diff
@@ -3,3 +3,4 @@
 build/
 node_modules/
 .env
+*.p8
```
Comment (young-k) [inline .gitignore]: what is a `.p8` file?

File: README.md [hunk 9]
```diff
@@ -2,3 +2,217 @@
...
+````
+If you get a database error, upon running `npm run start:dev` and you already have the database created try
+````
+DROP DATABASE push-notifications;
+CREATE DATABASE push-notifications;
+````
+Connect to the database by running
+````
+psql push-notifications
+````
```
Comment (young-k) [inline README.md]: nice instructions here

File: src/db/DBConnection.ts [hunk 10]
```diff
@@ -1,8 +1,12 @@
 import { Connection, ConnectionOptions, createConnection } from 'typeorm';
 
-const isProduction = process.env.NODE_ENV === 'production';
+import App from '../entities/App';
+import { PRODUCTION } from '../common/constants';
+import User from '../entities/User';
```
Comment (young-k) [inline src/db/DBConnection.ts]: alphabetize

File: src/repos/AppsRepo.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
+import { getConnectionManager, Repository } from 'typeorm';
+
+import App from '../entities/App';
+
+const db = (): Repository<App> => getConnectionManager().get().getRepository(App);
```
Comment (young-k) [inline src/repos/AppsRepo.ts]: should the `getConnectionManager.get()` be in a singleton-like structure if it's shared across all of the repos? it's a design decision, but it's probably cleaner if it is

File: src/repos/UsersRepo.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+): Promise<User> => {
+  try {
+    const user = await db().findOne({ uuid });
+    if (deviceToken !== undefined) {
+      user.deviceToken = deviceToken;
+      user.notificationsEnabled = true;
+    }
+    if (notificationsEnabled !== undefined) {
+      user.notificationsEnabled = notificationsEnabled;
+    }
```
Comment (young-k) [inline src/repos/UsersRepo.ts]: why isn't this snippet:

```javascript
user.deviceToken = ...
user.notificationsEnabled = ...
```

Rather than have these be optional parameters, there should just be default values, so we don't have to even check whether they're undefined or not! Optional parameters should be avoided in nearly any scenario, especially in a strongly typed language!

File: src/db/DBConnection.ts [hunk 13]
```diff
@@ -1,8 +1,12 @@
...
 
-const entities = [];
+const isProduction = process.env.NODE_ENV === PRODUCTION;
```
Comment (young-k) [inline src/db/DBConnection.ts]: why is this here if there's an `EnvUtils` file?

File: src/routers/UpdateUserRouter.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+import AppsRepo from '../repos/AppsRepo';
+import { SerializedUser } from '../common/types';
+import UsersRepo from '../repos/UsersRepo';
+
+class UpdateUserRouter extends ApplicationRouter<SerializedUser> {
+  constructor() {
+    super('POST');
+  }
+
+  getPath(): string {
```
Comment (young-k) [inline src/routers/UpdateUserRouter.ts]: is this type annotation necessary here or anywhere else, if it's extending the superclass? if it's unnecessary, we should remove it because it makes our code less DRY

File: src/entities/App.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
+import { Column, Entity, OneToMany } from 'typeorm';
+import { SerializedApp } from '../common/types';
+import Base from './Base';
```
Comment (young-k) [inline src/entities/App.ts]: sometimes you have a newline after the standard imports and sometimes you don't, pick one (i think having newlines is better), and stick to it!

File: src/entities/User.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+@Entity('users')
+class User extends Base {
+  @Column({ unique: true })
+  uuid: string;
+
+  @Column({ unique: true })
+  deviceToken: string;
+
+  @Column()
+  notificationsEnabled: boolean;
```
Comment (young-k) [inline src/entities/User.ts]: prefer that boolean values begin with `is`, so `isNotificationsEnabled`

File: src/db/DBConnection.ts (hunk 10 above)
Comment (kevinchan159) [reply src/db/DBConnection.ts]: do we alphabetized based on what we import or the filename?

**Comment by young-k** [comment]
Also, make sure you squash your commits!

File: src/db/DBConnection.ts (hunk 10 above)
Comment (young-k) [reply src/db/DBConnection.ts]: i believe @meganle wants by what's imported, but either way the `{ PRODUCTION }` belongs one above `App`

File: .gitignore (hunk 8 above)
Comment (kevinchan159) [reply .gitignore]: i explained it in README

Comment (young-k) [reply .gitignore]: prefer that we have a name for it then, `appdev.p8` or something –– `*.p8` implies that there are multiple

Comment (kevinchan159) [reply .gitignore]: the name of the file is typically `AuthKey_{APN_AUTH_KEY_ID}.p8`. i wanted to keep it generalizable to outside AppDev so i kept it like that. thoughts?

File: README.md [hunk 17]
```diff
@@ -2,3 +2,217 @@
...
+
+
+----------
+# **/users/update/** • POST
+
+**Description:** Updates information about a registered user.
+
+## Request Body
+    {
+      "deviceToken": String?,
```
Comment (meganle) [inline README.md]: ?String and ?Boolean

File: src/entities/User.ts [hunk 18]
```diff
@@ -0,0 +1,33 @@
+import { Column, Entity, OneToMany, ManyToOne } from 'typeorm';
```
Comment (meganle) [inline src/entities/User.ts]: ManyToOne, OneToMany

File: src/routers/UpdateUserRouter.ts [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+  getPath() {
+    return '/users/update/';
+  }
+
+  async content(req: Request): Promise<SerializedUser> {
+    const { deviceToken, isNotificationsEnabled, uuid } = req.body;
+
+    if (!uuid) throw Error('Missing required uuid');
+    if (deviceToken === undefined && isNotificationsEnabled === undefined) {
```
Comment (meganle) [inline src/routers/UpdateUserRouter.ts]: you could do `!deviceToken`

File: src/routers/UpdateUserRouter.ts [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+      throw Error('Either deviceToken or isNotificationsEnabled is required');
+    }
+
+    let user;
+    if (deviceToken !== undefined) {
```
Comment (meganle) [inline src/routers/UpdateUserRouter.ts]: `if (deviceToken)`
